src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs

```diff
@@ -495,26 +501,65 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
                     var elemRawType = vectorType.ItemType.RawType;
                     var srcNamedValueGetters = GetNamedOnnxValueGetters(input, _inputColIndices, _inputOnnxTypes, _inputTensorShapes);
                     if (vectorType.ItemType is TextDataViewType)
-                        return MakeStringTensorGetter(input, iinfo, srcNamedValueGetters, activeOutputColNames);
+                        return MakeStringTensorGetter(input, iinfo, srcNamedValueGetters, activeOutputColNames, outputCacher);
                     else
-                        return Utils.MarshalInvoke(MakeTensorGetter<int>, elemRawType, input, iinfo, srcNamedValueGetters, activeOutputColNames);
+                        return Utils.MarshalInvoke(MakeTensorGetter<int>, elemRawType, input, iinfo, srcNamedValueGetters, activeOutputColNames, outputCacher);
                 }
                 else
                 {
                     var type = _parent.Model.ModelInfo.OutputsInfo[_parent.MapDataViewColumnToOnnxOutputTensor(iinfo)].DataViewType.RawType;
                     var srcNamedValueGetters = GetNamedOnnxValueGetters(input, _inputColIndices, _inputOnnxTypes, _inputTensorShapes);
-                    return Utils.MarshalInvoke(MakeObjectGetter<int>, type, input, iinfo, srcNamedValueGetters, activeOutputColNames);
+                    return Utils.MarshalInvoke(MakeObjectGetter<int>, type, input, iinfo, srcNamedValueGetters, activeOutputColNames, outputCacher);
+                }
+            }
+
+            Delegate[] IRowMapper.CreateGetters(DataViewRow input, Func<int, bool> activeOutput, out Action disposer)
```

Comment:
Good point! I don't know why I didn't think of trying it. Let me try it and see. It will obviously a cleaner solution.

---
In reply to: [533779304](https://github.com/dotnet/machinelearning/pull/5518#discussion_r533779304) [](ancestors = 533779304)

Reply:
Looks like that would make it messier. MakeGetter is also declared in MapperBase as an abstract function and I will have to implement that. And the signature for that is not correct for what OnnxTransformer needs. I can have my version of CreateGetters not call MakeGetter, but then I still need to implement MakeGetter to satisfy the base class. That leaves it a bit inconsistent.

---
In reply to: [533781345](https://github.com/dotnet/machinelearning/pull/5518#discussion_r533781345) [](ancestors = 533781345,533779304)

Reply:
Makes sense. I've tried to think of an alternative to this, but this does seem to be the best solution to avoid refactoring more things. So I guess it's ok to leave it this way. I guess this is why OnnxTransformer used to have a Cacher for each column, but as you mentioned, this meant no actual caching was done. 😕

Reply:
> I can have my version of CreateGetters not call MakeGetter, but then I still need to implement MakeGetter to satisfy the base class.

You can just have your override of `MakeGetter` `throw new NotSupportedException("this should never be called.");`

Reply:
Good idea! I have implemented this and reverted back to inheriting from MapperBase

---
In reply to: [533817299](https://github.com/dotnet/machinelearning/pull/5518#discussion_r533817299) [](ancestors = 533817299)